### PR TITLE
Fix StructuredDict with nested JSON schemas using $ref

### LINKED ISSUE #2466: StructuredDict passed into run cannot handle nested JsonSchema
### Initial Checks

- [x] I confirm that I'm using the latest version of Pydantic AI
- [x] I confirm that I searched for my issue in https://github.com/pydantic/pydantic-ai/issues before opening this issue

### Description

```py
# %%
from pydantic import BaseModel
from pydantic_ai import Agent, StructuredDict


class Tire(BaseModel):
    brand: str


class Car(BaseModel):
    tires: list[Tire]


car_schema = Car.model_json_schema()


structured_dict = StructuredDict(car_schema)
print(structured_dict)
agent = Agent("gpt-4o-mini", output_type=structured_dict)
result = agent.run("What is the make of the car?")


# %%
import json
print(json.dumps(car_schema, indent=4))
```

```json
{
    "$defs": {
        "Tire": {
            "properties": {
                "brand": {
                    "title": "Brand",
                    "type": "string"
                }
            },
            "required": [
                "brand"
            ],
            "title": "Tire",
            "type": "object"
        }
    },
    "properties": {
        "tires": {
            "items": {
                "$ref": "#/$defs/Tire"
            },
            "title": "Tires",
            "type": "array"
        }
    },
    "required": [
        "tires"
    ],
    "title": "Car",
    "type": "object"
```

But `run()` crashes with 

```
KeyError: '#/$defs/Tire'
```

### Example Code

```Python

```

### Python, Pydantic AI & LLM client version

```Text
0.6.0, python 3.12
```

### PR DESCRIPTION
**Pull request by ChiaXinLiang**
## Summary

Fixes #2466 - StructuredDict now properly handles nested JSON schemas with `$ref` references.

## Problem

`StructuredDict` was failing when given JSON schemas containing nested `$ref` references because pydantic's JSON schema generator couldn't resolve them, resulting in a `KeyError`.

Example schema that was failing:
```python
{
    "$defs": {
        "Tire": {
            "type": "object",
            "properties": {
                "brand": {"type": "string"},
                "size": {"type": "integer"}
            }
        }
    },
    "type": "object",
    "properties": {
        "tires": {
            "type": "array",
            "items": {"$ref": "#/$defs/Tire"}
        }
    }
}
```

## Solution

1. **Added helper functions** to detect and resolve nested references:
   - `_contains_ref()`: Recursively checks if a schema contains any `$ref` references
   - `_resolve_refs()`: Recursively resolves all `$ref` references inline

2. **Updated `StructuredDict`** to:
   - Detect when a schema contains `$defs` with nested references
   - Resolve all references inline before passing to pydantic
   - Remove `$defs` after inlining to avoid confusion

3. **Improved `check_object_json_schema`**:
   - Replaced hacky string matching with proper recursive checking
   - Handles unresolvable refs gracefully (external URLs, missing refs, non-standard formats)

## Test plan

- [x] Added `test_output_type_structured_dict_nested` to test the main fix with nested car/tire schema
- [x] Added `test_output_type_structured_dict_unresolvable_ref` to test edge cases
- [x] Added comprehensive coverage tests in `test_utils.py`
- [x] All existing tests pass
- [x] Linting and formatting checks pass
- [x] Type checking with pyright passes

## Example Usage

```python
from pydantic_ai import Agent, StructuredDict

car_schema = {
    "$defs": {
        "Tire": {
            "type": "object",
            "properties": {
                "brand": {"type": "string"},
                "size": {"type": "integer"}
            }
        }
    },
    "type": "object",
    "properties": {
        "make": {"type": "string"},
        "model": {"type": "string"},
        "tires": {
            "type": "array",
            "items": {"$ref": "#/$defs/Tire"}
        }
    }
}

CarDict = StructuredDict(car_schema, name="Car")
agent = Agent('openai:gpt-4o', output_type=CarDict)
result = agent.run_sync('Create a car')
# Now works without KeyError\!
```

### CODE REVIEW

**Comment by DouweM** [comment]
@ChiaXinLiang Thanks, I agree that until this is fixed in Pydantic we should have a workaround. Could we use the existing [`InlineDefsJsonSchemaTransformer`](https://github.com/pydantic/pydantic-ai/blob/eeeb32d5a9d23efc52813e660fd6101c42b5c5b6/pydantic_ai_slim/pydantic_ai/profiles/_json_schema.py#L181) though?

File: tests/test_utils.py [hunk 1]
```diff
@@ -116,6 +117,53 @@ def test_check_object_json_schema():
...
+    assert check_object_json_schema(schema_with_external_ref) == schema_with_external_ref
+
+    # Test with different format ref (not starting with #/$defs/)
+    schema_with_different_ref = {
+        '$ref': '#/definitions/Model',
+    }
+    assert check_object_json_schema(schema_with_different_ref) == schema_with_different_ref
+
+
+def test_get_traceparent_coverage():
```
Comment (DouweM) [inline tests/test_utils.py]: This looks unrelated, can we remove it?

File: tests/test_agent.py [hunk 2]
```diff
@@ -1362,6 +1366,92 @@ def call_tool(_: list[ModelMessage], info: AgentInfo) -> ModelResponse:
...
+    NoRefsDict = StructuredDict(schema_no_refs, name='NoRefs')
+
+    # These should all work without errors even with unresolvable refs
+    assert MissingDict.__name__ == '_StructuredDict'
+    assert ExternalDict.__name__ == '_StructuredDict'
+    assert DifferentDict.__name__ == '_StructuredDict'
+    assert NoRefsDict.__name__ == '_StructuredDict'
+
+
+def test_structured_dict_name_description():
```
Comment (DouweM) [inline tests/test_agent.py]: This looks unrelated, can we remove it?

File: tests/test_agent.py [hunk 3]
```diff
@@ -1362,6 +1366,92 @@ def call_tool(_: list[ModelMessage], info: AgentInfo) -> ModelResponse:
...
+
+    # Schema with no refs at all (for coverage of line 347 in output.py)
+    schema_no_refs = {
+        'type': 'object',
+        'properties': {'simple': {'type': 'string'}},
+        '$defs': {'Unused': {'type': 'object', 'properties': {'name': {'type': 'string'}}}},
+    }
+    NoRefsDict = StructuredDict(schema_no_refs, name='NoRefs')
+
+    # These should all work without errors even with unresolvable refs
```
Comment (DouweM) [inline tests/test_agent.py]: This looks unrelated, can we remove it?

File: tests/test_agent.py [hunk 4]
```diff
@@ -1362,6 +1366,92 @@ def call_tool(_: list[ModelMessage], info: AgentInfo) -> ModelResponse:
...
+
+    # Schema with external ref
+    schema_with_external_ref = {'$ref': 'http://example.com/schemas/model.json'}
+    ExternalDict = StructuredDict(schema_with_external_ref, name='ExternalRef')
+
+    # Schema with non-standard ref format
+    schema_with_different_ref = {'$ref': '#/definitions/Model'}
+    DifferentDict = StructuredDict(schema_with_different_ref, name='DifferentRef')
+
+    # Schema with no refs at all (for coverage of line 347 in output.py)
```
Comment (DouweM) [inline tests/test_agent.py]: That line number will change over time, so no need to include it

File: tests/test_agent.py [hunk 5]
```diff
@@ -1362,6 +1366,92 @@ def call_tool(_: list[ModelMessage], info: AgentInfo) -> ModelResponse:
...
+    }
+    MissingDict = StructuredDict(schema_with_missing_ref, name='MissingRef')
+
+    # Schema with external ref
+    schema_with_external_ref = {'$ref': 'http://example.com/schemas/model.json'}
+    ExternalDict = StructuredDict(schema_with_external_ref, name='ExternalRef')
+
+    # Schema with non-standard ref format
+    schema_with_different_ref = {'$ref': '#/definitions/Model'}
+    DifferentDict = StructuredDict(schema_with_different_ref, name='DifferentRef')
```
Comment (DouweM) [inline tests/test_agent.py]: Should we verify that the schema is still as expected?

File: tests/test_agent.py [hunk 6]
```diff
@@ -1362,6 +1366,92 @@ def call_tool(_: list[ModelMessage], info: AgentInfo) -> ModelResponse:
...
+            'required': ['make', 'model', 'tires'],
+        },
+        name='Car',
+        description='A car with tires',
+    )
+
+    def call_tool(_: list[ModelMessage], info: AgentInfo) -> ModelResponse:
+        assert info.output_tools is not None
+        args_json = '{"make": "Toyota", "model": "Camry", "tires": [{"brand": "Michelin", "size": 17}]}'
+        return ModelResponse(parts=[ToolCallPart(info.output_tools[0].name, args_json)])
```
Comment (DouweM) [inline tests/test_agent.py]: Can we add an assertion for the output tool schema to ensure it has the expected format?

File: pydantic_ai_slim/pydantic_ai/_utils.py [hunk 7]
```diff
@@ -77,15 +77,40 @@ def check_object_json_schema(schema: JsonSchemaValue) -> ObjectJsonSchema:
...
+            defs = schema.get('$defs', {})
+            if ref_name in defs:
+                resolved = defs[ref_name]
+                # Check if the resolved schema contains nested references
+                if _contains_ref(resolved):
+                    # Keep the $defs because they're needed for nested references
+                    return schema
+                return resolved
+        # For non-local refs or unresolvable refs, return the schema as-is
+        return schema
```
Comment (DouweM) [inline pydantic_ai_slim/pydantic_ai/_utils.py:83]: Why was this change necessary? What was not working correctly before that is now?

**Comment by ChiaXinLiang** [comment]
Should I need to include the 'test_output_type_structured_dict_unresolvable_ref' also, or just 'test_output_type_structured_dict_nested'?

**Comment by ChiaXinLiang** [comment]
When I removed the `test_output_type_structured_dict_unresolvable_ref` test function. 

  The removed tests were covering these edge cases:
  - Non-local refs (e.g., http://example.com/schemas/model.json) - line 95 in _utils.py
  - Missing definition refs (e.g., ref to non-existent $defs) - line 95 in _utils.py
  - UserError exception handling in InlineDefsJsonSchemaTransformer - line 316 in output.py
  - List traversal in _contains_ref() function - line 114 in _utils.py

**Comment by DouweM** [comment]
@ChiaXinLiang Just a heads-up that I'll be out this coming week and will be back the 25th. Assuming this is not urgent I'll review it then. If it is, please ping `Kludex`! Appreciate the patience :)

**Comment by ChiaXinLiang** [comment]
Got it, thanks. No urgency on my end. Enjoy your week off!

File: pydantic_ai_slim/pydantic_ai/output.py [hunk 8]
```diff
@@ -304,6 +305,19 @@ def StructuredDict(
...
+    # JSON schema generator (Issue #2466)
+    if '$defs' in json_schema:
+        from .profiles import InlineDefsJsonSchemaTransformer
+
+        try:
+            transformer = InlineDefsJsonSchemaTransformer(json_schema)
+            json_schema = transformer.walk()
+        except UserError:
+            # If the transformer can't resolve refs (e.g., missing definitions),
+            # keep the original schema unchanged
```
Comment (DouweM) [inline pydantic_ai_slim/pydantic_ai/output.py]: Let's raise a descriptive error instead of continuing and hitting the confusing Pydantic error

**Review by DouweM — CHANGES_REQUESTED** [review]

**Comment by DouweM** [comment]
@ChiaXinLiang Please have a look at the remaining unresolved comments from 2 weeks ago and the new one!

**Comment by DouweM** [comment]
@ChiaXinLiang Thanks for your work on this! I found some time to finish it up :)
